Nonprofit-Exchange-Hub/web-app
# Validate organizations

## Dev Summary

This PR:
**1. Creates an endpoint for the EIN check in the Organizations Controller** (GET "/organizations/ein/:ein"). 
This endpoint will return a ProPublica organization object or `null`. 

An abbreviated valid EIN return example: 
```
{
    "id": 142007220,
    "ein": 142007220,
    "name": "PRO PUBLICA INC",
    "careofname": null,
    "address": "155 AVE AMERICA 13 FL",
    "city": "NEW YORK",
    "state": "NY",
    "zipcode": "10013-0000",
    "exemption_number": 0,
    "subsection_code": 3,
    "affiliation_code": 3,
    "classification_codes": "1000",
    "ruling_date": "2008-02-01",
    "deductibility_code": 1,
    "foundation_code": 15,
 ...
}
```

This will allow a request to be made from the front end to verify the organization prior to a user being allowed to save the instance. 

The object will give the frontend access not only to the EIN and name, but perhaps other helpful information for the future. 

**2. This PR also adds the EINCheck function in the Organizations Service** for the new endpoint to utilize. This makes a fetch request, checking ProPublica's API for that EIN number, returning that organization as saved in their API. 

Originally, I attempted to make the API call using the recommended Axios module per the [Nest documentation](https://docs.nestjs.com/techniques/http-module). James and I spent a lot of time barking up this tree, trying to utilize Observables and struggling with the asynchronous calls. Ultimately, he recommended I use `fetch` instead, which turned out to be a wildly simpler implementation. This way I was able to return an object wrapped in an easy to understand Promise, rather than a more confusing, Observable. 

Furthermore, I originally incorporated the name and EIN cross-checking logic in the service, upon the organization's creation. However, upon review and discussion with James regarding the user experience, we decided to save whatever was sent from the front end in the ```create``` function, and to more simply return the org object from the ```EINCheck``` function. 

## Test Plan
To test this PR, make get requests to the EIN endpoint with both valid, and invalid EINs. 

preReqs (to be checked by reviewer):

- [ ] following command succeeds:
`cd client && npm run checks && cd ../server && npm run checks && cd ..`
expect tests to pass
- [ ] ensure all files have been formatted:
`cd client && npm run format && cd ../server && npm run format && cd ..`
expect `git status` to show 0 files changed


repro steps:

1. Make a GET request via Postman or other http service to: `http://localhost:3001/api/organizations/ein/142007220` With this valid EIN, you should receive an object with many organization fields: 

<img width="622" alt="PPObj" src="https://user-images.githubusercontent.com/76884037/148290129-2fefea27-6235-4d9f-9eb2-573c59d12443.png">

2. Make a GET request with an invalid EIN. A la: `http://localhost:3001/api/organizations/ein/242007220` With this invalid EIN, you should receive nothing: 

<img width="867" alt="null" src="https://user-images.githubusercontent.com/76884037/148290114-5a136263-9ee9-411b-a731-7ca5384632d1.png">



## Resources
- [Propublica API](https://projects.propublica.org/nonprofits/api/)
- [utilizing exceptions](https://docs.nestjs.com/exception-filters)
- [GET request with params](https://stackoverflow.com/questions/54958244/how-to-use-query-parameters-in-nest-js)
- [making fetch requests in node](https://blog.logrocket.com/5-ways-to-make-http-requests-in-node-js/)
